python/pyiceberg/io/fsspec.py

```diff
@@ -0,0 +1,182 @@
+
+    Args:
+        location(str): A URI to a file location
+        fs: An fsspec filesystem instance
+    """
+
+    def __init__(self, location: str, fs):
+        self._fs = fs
+        super().__init__(location=location)
+
+    def __len__(self) -> int:
+        """Returns the total length of the file, in bytes"""
+        object_info = self._fs.info(self.location)
+        if size := object_info.get("Size"):
+            return size
+        elif size := object_info.get("size"):
+            return size
+        raise RuntimeError(f"Cannot retrieve object info: {self.location}")
+
+    def exists(self) -> bool:
+        """Checks whether the location exists"""
+        return self._fs.lexists(self.location)
+
+    def create(self, overwrite: bool = False):
+        """Create an output stream for reading the contents of the file
+
+        Args:
+            overwrite(bool): Whether to overwrite the file if it already exists
+
+        Returns:
+            OpenFile: An fsspec compliant file-like object
+
+        Raises:
+            FileExistsError: If the file already exists at the location and overwrite is set to False
+
+        Note:
+            If overwrite is set to False, a check is first performed to verify that the file does not exist.
+            This is not thread-safe and a possibility does exist that the file can be created by a concurrent
+            process after the existence check yet before the output stream is created. In such a case, the default
+            behavior will truncate the contents of the existing file when opening the output stream.
+        """
+        if not overwrite and self.exists():
+            raise FileExistsError(f"Cannot create file, file already exists: {self.location}")
+        self._fs.touch(self.location)
+        return self._fs.open(self.location, "wb")
+
+    def to_input_file(self) -> FsspecInputFile:
+        """Returns a new FsspecInputFile for the location at `self.location`"""
+        return FsspecInputFile(location=self.location, fs=self._fs)
+
+
+class FsspecFileIO(FileIO):
+    """A FileIO implementation that uses fsspec"""
+
+    def __init__(self, client_kwargs: Dict[str, str] = EMPTY_DICT, **fs_kwargs):
+        client_properties = {f"client_{key}": val for key, val in client_kwargs.items()}
+        fs_properties = {f"fs_{key}": val for key, val in fs_kwargs.items()}
+        properties = {**client_properties, **fs_properties}
+        self._fs = None
+        super().__init__(properties=properties)
+
+    def new_input(self, location: str) -> FsspecInputFile:
+        """Get an FsspecInputFile instance to read bytes from the file at the given location
+
+        Args:
+            location(str): A URI or a path to a local file
+
+        Returns:
+            FsspecInputFile: An FsspecInputFile instance for the given location
+        """
+        uri = urlparse(location)
+        fs = self.properties.get("fs") or self._resolve_fs_type(scheme=uri.scheme)
```

Comment:
I don't think it should be possible to pass in a specific FS. Callers should register a method to create a FS (even if that just returns a configured one). I'm not sure if that should be done in the constructor or not, since we are passing properties as `kwargs`.

@Fokko, should we pass properties as `properties: Properties` rather than `**properties: str` so that we can pass other optional arguments?

Reply:
`properties: Properties` is what the parent class `FileIO` uses. I used that here and it worked nicely. @Fokko let me know if that works.

Reply:
If we don't want to mix certain properties, then we should pass in different dicts. I like how we have it now. We could even use TypedDicts in the future if we want to be more explicit: https://peps.python.org/pep-0589/